### PR TITLE
Set Client.as_current when entering ctx

### PR DESCRIPTION
**Pull request by fjetter**
Most of the time we actually prefer using the current client and not the default client. Whenever we're using the client in a ctx manager, we should set the current client

TODO: tests

cc @graingert

### CODE REVIEW

File: distributed/client.py [hunk 1]
```diff
@@ -1444,11 +1444,13 @@ def _heartbeat(self):
     def __enter__(self):
         if not self._loop_runner.is_started():
             self.start()
-        return self
+        with self.as_current():
+            return self
```
[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/client.py:1448]: And then we want to hold onto the state that is returned, and track that state and reuse it in the `__exit__` call.

[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/client.py:1448]: yes, you are right of course.

[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/client.py:1448]: I briefly tried this but ran into `Client._close` not being awaited.  Not sure why.

```diff
diff --git a/distributed/client.py b/distributed/client.py
index c3952ec0..19cf81e7 100644
--- a/distributed/client.py
+++ b/distributed/client.py
@@ -1444,10 +1444,12 @@ class Client(SyncMethodMixin):
     def __enter__(self):
         if not self._loop_runner.is_started():
             self.start()
+        self._previous_as_current = _current_client.set(self)
         return self

     async def __aenter__(self):
         await self
+        self._previous_as_current = _current_client.set(self)
         return self

     async def __aexit__(self, exc_type, exc_value, traceback):
@@ -1457,9 +1459,11 @@ class Client(SyncMethodMixin):
             fast=exc_type
             is not None
         )
+        _current_client.reset(self._previous_as_current)

     def __exit__(self, exc_type, exc_value, traceback):
         self.close()
+        _current_client.reset(self._previous_as_current)

     def __del__(self):
         # If the loop never got assigned, we failed early in the constructor,
diff --git a/distributed/tests/test_client.py b/distributed/tests/test_client.py
index 3ee6c779..748c6150 100644
--- a/distributed/tests/test_client.py
+++ b/distributed/tests/test_client.py
@@ -4074,10 +4074,10 @@ async def test_as_current(c, s, a, b):
             assert Client.current() is c
             with pytest.raises(ValueError):
                 Client.current(allow_global=False)
-            with c1.as_current():
+            with c1:
                 assert Client.current() is c1
                 assert Client.current(allow_global=True) is c1
-            with c2.as_current():
+            with c2:
                 assert Client.current() is c2
                 assert Client.current(allow_global=True) is c2
```

File: distributed/tests/test_client.py [hunk 2]
```diff
@@ -3365,16 +3365,20 @@ def test_default_get(loop_in_thread):
...
 
-@gen_cluster(client=True)
-async def test_ensure_default_client(c, s, a, b):
-    assert c is default_client()
-
-    async with Client(s.address, set_as_default=False, asynchronous=True) as c2:
+@gen_cluster()
+async def test_ensure_default_client(s, a, b):
+    c = await Client(s.address, asynchronous=True)
+    try:
```
Comment (fjetter) [inline distributed/tests/test_client.py:3371]: Turns out we can only set current when entering ctx when `set_as_default` is also set. Otherwise this breaks tests like this.

That also makes sense considering that `default_client` and `Client.current` are effectively aliases

File: distributed/tests/test_client.py [hunk 3]
```diff
@@ -4072,8 +4076,6 @@ async def test_as_current(c, s, a, b):
 
         with temp_default_client(c):
             assert Client.current() is c
-            with pytest.raises(ValueError):
-                Client.current(allow_global=False)
             with c1.as_current():
```
Comment (mrocklin) [inline distributed/tests/test_client.py:4079]: In tests I was playing with removing `as_current`.  Presumably `with client.as_current` should now be equivalent to `with client` and shouldn't be something that we recommend any longer.

Comment (fjetter) [reply distributed/tests/test_client.py:4079]: Interesting problem. When I did this I encountered the `Client._close` not awaited problem you mentioned before.

This is introduced (in my case) because I switched `temp_default_client` to use `with client` instead of `with client.as_current()`.

Now, whenever this temp_default_client exits the context, it calls `__exit__` even on asynchronous clients.

https://github.com/dask/distributed/blob/a9c240156544c99a58e6a6925ae34db7ffbc1096/distributed/client.py#L1709-L1713

here we are creating the close task but are never awaiting them.

TLDR `with client` and `with client.as_current()` is not equivalent since `with client` is closing the client as well.

Comment (fjetter) [reply distributed/tests/test_client.py:4079]: Interesting problem. When I did this I encountered the `Client._close` not awaited problem you mentioned before.

This is introduced (in my case) because I switched `temp_default_client` to use `with client` instead of `with client.as_current()`.

Now, whenever this temp_default_client exits the context, it calls `__exit__` even on asynchronous clients.

https://github.com/dask/distributed/blob/a9c240156544c99a58e6a6925ae34db7ffbc1096/distributed/client.py#L1709-L1713

here we are creating the close task but are never awaiting them.

TLDR `with client` and `with client.as_current()` is not equivalent since `with client` is closing the client as well.

File: distributed/tests/test_client.py [hunk 4]
```diff
@@ -4022,8 +4026,6 @@ async def test_as_current(c, s, a, b):
     ) as c2:
         with temp_default_client(c):
             assert Client.current() is c
```
Comment (graingert) [inline distributed/tests/test_client.py:4028]:
```suggestion
            assert Client.current() is c
            assert Client.current(allow_global=False) is c
```

File: distributed/tests/test_client.py [hunk 5]
```diff
@@ -3294,16 +3410,20 @@ async def test_get_scheduler_default_client_config_interleaving(s):
...
 
-@gen_cluster(client=True)
-async def test_ensure_default_client(c, s, a, b):
-    assert c is default_client()
-
-    async with Client(s.address, set_as_default=False, asynchronous=True) as c2:
+@gen_cluster()
+async def test_ensure_default_client(s, a, b):
+    c = await Client(s.address, asynchronous=True)
+    try:
```
Comment (crusaderky) [inline distributed/tests/test_client.py:3416]:
```suggestion
    async with Client(s.address, asynchronous=True) as c:
```

File: distributed/tests/test_client.py [hunk 6]
```diff
@@ -3294,16 +3410,20 @@ async def test_get_scheduler_default_client_config_interleaving(s):
...
-        assert c2 is default_client()
+
+        async with Client(s.address, set_as_default=False, asynchronous=True) as c2:
+            assert c is default_client()
+            assert c2 is not default_client()
+            ensure_default_client(c2)
+            assert c is not default_client()
+            assert c2 is default_client()
+    finally:
+        await c.close()
```
Comment (crusaderky) [inline distributed/tests/test_client.py:3426]:
```suggestion
```

**Review by crusaderky — APPROVED** [review]
Just a trivial nitpick
